kb/development/test_suites.md

```diff
@@ -114,3 +114,112 @@ void test_function_example( char *input, char *expected_output, int expected_ret
+
+In a test case that uses PSA crypto only when building with `MBEDTLS_USE_PSA_CRYPTO`, call `USE_PSA_INIT()` at the beginning and `USE_PSA_DONE()` at the end.
+
+See [`tests/include/test/psa_crypto_helpers.h`](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/include/test/macros.h) for more complex cases.
+
+## Guidance on writing unit test data
+
+### Document the test data
+
+Document how the test data was generated. This helps future maintainers if they need to generate more similar test data, for example to construct a non-regression test for a bug in a particular case.
+
+The documentation can be:
+
+* In Python (or other) code committed to the repository. This is preferred when the code can handle a large class of tests. For example, we have frameworks to generate key data for PSA, and to generate bignum tests. Some `.data` files are fully generated at build time by `tests/scripts/generate_*_tests.py`.
+* In a comment in the `.data` file. This is more convenient when the instructions are specific to a single test case.
+* In the commit that introduces the test data. This is more convenient when the instructions cover a series of similar test cases.
+
+When adding a new file in `tests/data_files`, if possible, do it by editing `tests/data_files/Makefile` to add executable instructions for creating those files, and then run `make` and commit the result. Ideally removing the files and re-running `make` should produce identical files, however in many cases this is not practical because the generation is randomized or depends on the current time (for certificates and related data), and we accept that.
+
+### Interoperability testing
+
+Test data for cryptographic algorithms should be, in order of preference:
+
+* Official test vectors taken from a standards document.
+* Generated by a reference implementation from the authors of the algorithm.
+* Generated by a widespread implementation such as OpenSSL or Cryptodome.
+* Generated by an independent implementation.
+* Obtained through manual means, possibly by patching together bits of other tests. For example, test output multipart operation function can be obtained by splitting the output of a one-shot operation function. Edge cases for parsing can be constructed by manually tweaking nominal cases.
+* As a last resort, obtained by running the library once. This is a last resort since it cannot validate that the implementation is correct, it can only protect against future behavior changes. This should only be done when there is no other way, for example to construct a non-regression test in an edge case if we're very confident from code review that our bug fix is correct.
+
+Always [document how the test data was constructed](#document-the-test-data).
```

Comment:
Is it necessary to repeat this given it's actually a section heading less than a page above? Not sure how it counts as "interoperability testing"

Reply:
I think it is useful to emphasise that it needs to be documented and recording the source of the test vectors is even more important here.

Reply:
I meant “whichever of these options you choose, remember to write down what you did”. I've reworded the sentence to actually say that.